rasa/utils/tensorflow/models.py

```diff
@@ -320,226 +170,64 @@ def train_on_batch(
-        call_model_function: Callable,
-        batch_size: int,
-        training: bool,
-        offset: int,
-        writer: Optional["ResourceSummaryWriter"] = None,
-    ) -> int:
-        """Run on batches."""
-        self.reset_metrics()
-
-        step = offset
-
-        self._training = training  # needed for eager mode
-        for batch_in in dataset_function(batch_size):
-            call_model_function(batch_in)
-
-            if not self.tensorboard_log_on_epochs:
-                self._log_metrics_for_tensorboard(step, writer)
-
-            step += 1
-
-        return step
-
-    @staticmethod
-    def _get_tf_call_model_function(
-        dataset_function: Callable,
-        call_model_function: Callable,
-        eager: bool,
-        phase: Text,
-    ) -> Callable:
-        """Convert functions to tensorflow functions."""
-        if eager:
-            return call_model_function
-
-        logger.debug(f"Building tensorflow {phase} graph...")
-
-        init_dataset = dataset_function(1)
-        tf_call_model_function = tf.function(
-            call_model_function, input_signature=[init_dataset.element_spec]
-        )
-        tf_call_model_function(next(iter(init_dataset)))
-
-        logger.debug(f"Finished building tensorflow {phase} graph.")
-
-        return tf_call_model_function
-
-    def _get_tf_train_functions(
-        self, eager: bool, model_data: RasaModelData, batch_strategy: Text
-    ) -> Tuple[Callable, Callable]:
-        """Create train tensorflow functions."""
-
-        def train_dataset_function(_batch_size: int) -> tf.data.Dataset:
-            return model_data.as_tf_dataset(_batch_size, batch_strategy, shuffle=True)
-
-        self._training = True  # needed for tf graph mode
-        return (
-            train_dataset_function,
-            self._get_tf_call_model_function(
-                train_dataset_function, self.train_on_batch, eager, "train"
-            ),
-        )
-
-    def _get_tf_evaluation_functions(
-        self, eager: bool, evaluation_model_data: Optional[RasaModelData]
-    ) -> Tuple[Optional[Callable], Optional[Callable]]:
-        """Create evaluation tensorflow functions."""
-        if evaluation_model_data is None:
-            return None, None
-
-        def evaluation_dataset_function(_batch_size: int) -> tf.data.Dataset:
-            return evaluation_model_data.as_tf_dataset(
-                _batch_size, SEQUENCE, shuffle=False
-            )
+        return self._get_metric_results()
 
-        self._training = False  # needed for tf graph mode
-        return (
-            evaluation_dataset_function,
-            self._get_tf_call_model_function(
-                evaluation_dataset_function, self._total_batch_loss, eager, "evaluation"
-            ),
-        )
+    def predict_step(
+        self, batch_in: Union[Tuple[tf.Tensor], Tuple[np.ndarray]]
+    ) -> Dict[Text, tf.Tensor]:
+        """Predict on batch."""
+        self.prepare_for_predict()
```

Comment:
`rasa shell` does not work. It has to be related to this line. Strangely `rasa shell nlu` works just fine. Unfortunately the error is not really helpful:
```
InvalidArgumentError: Unsuccessful TensorSliceReader constructor: Failed to get matching files on /var/folders/xk/kzf9l4h53k759yd46hv6qz140000gn/T/tmp9qwd4y25/nlu/component_7_DIETClassifier.tf_model: Not found: /var/folders/xk/kzf9l4h53k759yd46hv6qz140000gn/T/tmp9qwd4y25/nlu; No such file or directory [Op:RestoreV2]
```

Reply:
does returning to old style `if is not None` work?
my hunch is that `self.prepare_for_predict()` should be part of `build` but only when it is compiled for prediction

Reply:
or maybe part of `compile` not sure what is the difference to `build`

Reply:
> does returning to old style if is not None work?

No 😞 

As far as I understand `compile` is just needed before training. But need to take a closer look at all those methods again on Monday.

Reply:
Issue seems to be resolved. Just merged in `e2e`. Not sure exactly why it is now working, but maybe it was related to the way how the interpreter was load inside core.

Reply:
(Was on the wrong branch before - still not working.)

After further investigation I think the error isn't related to `self.prepare_for_predict()`. `rasa shell nlu` works fine. Also, when you parse a method as soon as the `Interpreter` is created inside `rasa shell` everything works just fine. However, as soon as the server is up the path to the model file seems to be not valid anymore. We need to check why the model file is needed once the model is already loaded.

Reply:
fixed it, we needed to keep the it on one example during loading to actually build the model and load the weights